File: javascript/yagni.js
// YAGNI (You Ain't Gonna Need It)
// Don't build a cannon to kill a fly; you might not even need it afterward.

// Before
class User {
  constructor(name, email) {
    this.name = name;
    this.email = email;
    this.age = null;  // Unused feature
    this.address = null;  // Unused feature
    this.phoneNumber = null;  // Unused feature
  }

  sendWelcomeEmail() {
    console.log(`Welcome, ${this.name}! A welcome email has been sent to ${this.email}.`);
  }
}

const user = new User("Alice", "alice@example.com");
user.sendWelcomeEmail();


// After
class User {
  constructor(name, email) {
    this.name = name;
    this.email = email;
  }

  sendWelcomeEmail() {
    console.log(`Welcome, ${this.name}! A welcome email has been sent to ${this.email}.`);
  }
}

const user = new User("Alice", "alice@example.com");
user.sendWelcomeEmail();

File: ruby/yagni.rb
# YAGNI (You Ain't Gonna Need It)
# Don't build a cannon to kill a fly; you might not even need it afterward.

# Before
class User
  attr_accessor :name, :email, :age, :address, :phone_number

  def initialize(name, email)
    @name = name
    @email = email
    @age = nil  # Unused feature
    @address = nil  # Unused feature
    @phone_number = nil  # Unused feature
  end

  def send_welcome_email
    puts "Welcome, #{name}! A welcome email has been sent to #{email}."
  end
end

user = User.new("Alice", "alice@example.com")
user.send_welcome_email


# After
class User
  attr_accessor :name, :email

  def initialize(name, email)
    @name = name
    @email = email
  end

  def send_welcome_email
    puts "Welcome, #{name}! A welcome email has been sent to #{email}."
  end
end

user = User.new("Alice", "alice@example.com")
user.send_welcome_email
